vignettes/basic_principles.Rmd

```diff
@@ -0,0 +1,111 @@
+---
+title: "Basic principles"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Basic principles}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+```{r setup}
+library(SCDB)
+options("SCDB.log_path" = tempdir())
+```
+
+The basic principle of the SCDB package is to enable the user to easily implement and maintain a data base of time-versioned data.
+
+In practice, this is done by labeling each record in the data with three additional information:
+
+* checksum: the md5 hash of the information stored in the record
+* from_ts: the timestamp at which the record was first introduced to the dataset
+* until_ts: the timestamp at which the record was first removed from the dataset
+
+Note: This strategy of time versioning is often called "type 2" history.
+
+
+The SCDB package provides the function `update_snapshot` to handle the insertion and deactivation of records using this strategy, and it further includes several functions to improve the Quality-of-life for working with data base data.
+
+Below we illustrate how this looks in practice:
+
+```{r Example}
+# First we connect to our data base.
+# If this is, e.g., a PostgreSQL data base already running on the machine, connection
+# can be done after the configuration of a .pgpass file.
+# For this example, we use an on-disk SQLite db to showcase.
+conn <- get_connection(drv = RSQLite::SQLite())
+# NOTE: Had the PostgreSQL DB been configured, we would not need to pass any args to get_connection()
+
+# Our example data is mtcars with rownames converted to a column and only the first hp column of mtcars
+example_data <- dplyr::transmute(mtcars, car = rownames(mtcars), hp)
+# If the data does not already live on the remote, we must transfer it
+example_data <- dplyr::copy_to(conn, example_data, overwrite = TRUE)
+
+# In this example, we imagine that on day 1, in this case 2020-01-01 11:00:00, our data known to us is
+# the first 3 records of mtcars
+data <- head(example_data, 3)
+
+# We then store these data in the data base using update_snapshot
+invisible(capture.output({  # NOTE: update_snapshot does some printing to terminal
+  update_snapshot(.data = data,
+                  conn = conn,
+                  db_table = "mtcars", # the name of the DB table we want to store the data in
+                  timestamp = as.POSIXct("2020-01-01 11:00:00"))
+}))
+
+# We can access our data using the `get_table` function
+print(get_table(conn, "mtcars"))
+
+# And we can see the time-keeping if we set `include_slice_info = TRUE`
+print(get_table(conn, "mtcars", include_slice_info = TRUE))
+
+
+
+# Lets say that the next day, our data set is now the first 5 of our example data
+data <- head(example_data, 5)
+
+# We then store these data in the data base using update_snapshot
+invisible(capture.output({
```

Comment:
Consider splitting the example into several code chunks and use the `eval = FALSE` option to a chunk instead of `invisible(capture.output(...`

Reply:
This is now done